Fix MRO problems with interface hierarchies

justussc opened the pull request:
The test in mro.py demonstrates the problem leading to the classic mro error. In larger interface hierarchies this problem can be hidden far down in the hierarchy such that one cannot only drop duplicates from the immedate bases as was done until now. Using a metaclass the mro is replaced by a simple topsort of the hierarchy.

Quite likely also Issue #159 is fixed by this change. 

(its my first step on github so please advice if i should have done something different)

marscher commented:
Thank you for you contribution!
Can you please write a single comment in the code on how you are calculating the order? (and if you're willing to touch this again, please remove the semicolons, Python is not C) :+1:

justussc commented:
i removed the semicolons (a habit from to much java/c++ coding). 
i also added a comment at the top of the _mro_override_topsort  function. as there is no true multiple inheritance in java, any linear ordering of the inheritance graph should be ok. 

following a recommendation on stackoverflow i ammended my commit and force pushed it. let me know if that is not the way to do this.

marscher commented:
Thank you again!
